### PR TITLE
Fix: Multiple errors in fibonacci search.

### PR DESCRIPTION
**Pull request by NumberPiOso**
- Test lists were not ordered, this is required for Fibonacci search
- Place documentation of function inside function
- Create multiple different tests including, float, char and negatives
- Add type hints in line with #2128

### **Describe your change:**

* [X] Fix a bug or typo in an existing algorithm?

### **Checklist:**
* [X] I have read [CONTRIBUTING.md](https://github.com/TheAlgorithms/Python/blob/master/CONTRIBUTING.md).
* [X] This pull request is all my own work -- I have not plagiarized.
* [X] I know that pull requests will not be merged if they fail the automated tests.
* [X] This PR only changes one algorithm file.  To ease review, please open separate PRs for separate algorithms.
* [X] All new Python files are placed inside an existing directory.
* [X] All filenames are in all lowercase characters with no spaces or dashes.
* [X] All functions and variable names follow Python naming conventions.
* [X] All function parameters and return values are annotated with Python [type hints](https://docs.python.org/3/library/typing.html).
* [X] All functions have [doctests](https://docs.python.org/3/library/doctest.html) that pass the automated testing.
* [X] All new algorithms have a URL in its comments that points to Wikipedia or other similar explanation.
* [X] If this pull request resolves one or more open issues then the commit message contains `Fixes: #{$ISSUE_NO}`.

### CODE REVIEW

File: searches/fibonacci_search.py [hunk 1]
```diff
@@ -1,51 +1,100 @@
...
+
+
+@lru_cache
+def fibonacci(k: int) -> int:
+    """Finds fibonacci number in index k.
 
+    Parameters
+    ----------
+    k : int
+        Index of fibonacci.
```
Comment (dhruvmanila) [inline searches/fibonacci_search.py:24]:
```suggestion
    Parameters
    ----------
    k : Index of fibonacci.
```
Already mentioned in type hints.

File: searches/fibonacci_search.py [hunk 2]
```diff
@@ -1,51 +1,100 @@
...
 
-def fibonacci_search(arr, val):
+    Returns
+    -------
+    int
+        Fibonacci number in position k.
```
Comment (dhruvmanila) [inline searches/fibonacci_search.py:29]:
```suggestion
    Returns
    -------
    Fibonacci number in position k.
```
Same as above

File: searches/fibonacci_search.py [hunk 3]
```diff
@@ -1,51 +1,100 @@
...
+    >>> print(fibonacci(2))
+    1
+    >>> print(fibonacci(5))
+    5
+    >>> print(fibonacci(15))
+    610
     """
-    >>> fibonacci_search([1,6,7,0,0,0], 6)
+    if k == 0:
+        return 0
```
Comment (dhruvmanila) [inline searches/fibonacci_search.py:41]: Please check for whether the input is `int` and greater than 0 and write an appropriate `doctest` for it.

File: searches/fibonacci_search.py [hunk 4]
```diff
@@ -1,51 +1,100 @@
...
+    arr : list
+        List of sorted elements.
+    val : int
+        Element to search in list.
+
+    Returns
+    -------
+    int
+        The index of the element in the array.
+        -1 if the element is not found.
```
Comment (dhruvmanila) [inline searches/fibonacci_search.py:62]:
```suggestion
    Parameters
    ----------
    arr : List of sorted elements.
    val : Element to search in the list.

    Returns
    -------
    The index of the element in the array.
    -1 if the element is not found.
```
As mentioned in the above comment.

File: searches/fibonacci_search.py [hunk 5]
```diff
@@ -1,51 +1,100 @@
...
+    n = len(arr)
+    # Find m such that F_m >= n where F_i is the i_th fibonacci number.
+    m = next(x for x in range(sys.maxsize ** 10) if fibonacci(x) >= n)
+    k = m
+    offset = 0
+    while k != 0:
+        if arr[offset + fibonacci(k - 1)] == val:
+            return fibonacci(k - 1)
+        elif val < arr[offset + fibonacci(k - 1)]:
+            k = k - 1
```
Comment (dhruvmanila) [inline searches/fibonacci_search.py:92]:
```suggestion
            k -= 1
```

File: searches/fibonacci_search.py [hunk 6]
```diff
@@ -1,51 +1,100 @@
...
+        elif val > arr[offset + fibonacci(k - 1)]:
+            offset += fibonacci(k - 2)
+            k = k - 2
```
Comment (dhruvmanila) [inline searches/fibonacci_search.py:95]:
```suggestion
            k -= 2
```

**Review by dhruvmanila — CHANGES_REQUESTED** [review]

File: searches/fibonacci_search.py (hunk 2 above)
Comment (NumberPiOso) [reply searches/fibonacci_search.py:29]: I am using NumPy Style Python Docstrings, and it is said that the return type must be duplicated in the docstring to comply.

URL
https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_numpy.html

@dhruvmanila why do you thinks it is better this other way?

File: searches/fibonacci_search.py [hunk 7]
```diff
@@ -35,7 +35,17 @@ def fibonacci(k: int) -> int:
...
+    Traceback (most recent call last):
+    ValueError: k must be an integer.
+    >>> print(fibonacci(-5))
+    Traceback (most recent call last):
+    ValueError: k integer must be greater or equal to zero.
     """
+    if not isinstance(k, int):
+        raise ValueError("k must be an integer.")
+    if k < 0:
+        raise ValueError("k integer must be greater or equal to zero.")
```
Comment (NumberPiOso) [inline searches/fibonacci_search.py:49]: I have a question in this one.

Do you think is it better to use Assertions or this conditionals?

File: searches/fibonacci_search.py (hunk 2 above)
Comment (dhruvmanila) [reply searches/fibonacci_search.py:29]:
```python
>>> import this
...
There should be one-- and preferably only one --obvious way to do it.
...
```

No duplications.

I have no idea about the NumPy Style Python Docstrings but if it is clear to read then it's fine.

File: searches/fibonacci_search.py (hunk 7 above)
Comment (dhruvmanila) [reply searches/fibonacci_search.py:49]: Assertions will raise `AssertionError` which is ambiguous, the user has no idea what the function is asserting. By raising specific errors, users can understand and the code is easier to read.

You can use `try ... except ...` block but that would become cumbersome and unnecessary for such small cases.

**Comment by NumberPiOso** [comment]
I think we are done here. Are not we?

File: searches/fibonacci_search.py [hunk 8]
```diff
@@ -1,51 +1,110 @@
...
+    >>> print(fibonacci('a'))
+    Traceback (most recent call last):
+    ValueError: k must be an integer.
+    >>> print(fibonacci(-5))
+    Traceback (most recent call last):
+    ValueError: k integer must be greater or equal to zero.
     """
-    >>> fibonacci_search([1,6,7,0,0,0], 6)
+    if not isinstance(k, int):
+        raise ValueError("k must be an integer.")
```
Comment (dhruvmanila) [inline searches/fibonacci_search.py:47]:
```suggestion
        raise TypeError("k must be an integer.")
```
As the user entered the wrong type.

File: searches/fibonacci_search.py [hunk 9]
```diff
@@ -1,51 +1,110 @@
...
-            fib_N_1 = fib_N_1 - fib_N_2
-            fib_N_2 = fibNext - fib_N_1
-        else:
-            return i
-    if (fib_N_1 and index < length - 1) and (arr[index + 1] == val):
-        return index + 1
-    return -1
+    n = len(arr)
+    # Find m such that F_m >= n where F_i is the i_th fibonacci number.
+    m = next(x for x in range(sys.maxsize ** 10) if fibonacci(x) >= n)
```
Comment (dhruvmanila) [inline searches/fibonacci_search.py:95]: `sys.maxsize` is already a huge number. Can you tell me why is this necessary?
And please use descriptive variable names instead of the single letter names as mentioned in CONTRIBUTING.md. Apply the changes wherever the variables were used.

**Review by dhruvmanila — CHANGES_REQUESTED** [review]

File: searches/fibonacci_search.py (hunk 9 above)
Comment (NumberPiOso) [reply searches/fibonacci_search.py:95]: `sys.maxsize` is a simple way to represent that the generator produced can go up to infinity. Probably is not the best way to find it.

It is equal to do something like this.

```
i = 0
while True:
    if fibbonacci(i) >= n:
        m = i
        break
    i += 1
```

Both are loops that can go up to infinity. But I do not know which one of them is clearer.

Comment (NumberPiOso) [reply searches/fibonacci_search.py]: I commented it inside the conflict, but I will paste the message here, because I think there it is not easy to find.

`sys.maxsize` is a simple way to represent that the generator produced can go up to infinity. Probably is not the best way to find it.

It is equal to do something like this.

```
i = 0
while True:
    if fibbonacci(i) >= n:
        m = i
        break
    i += 1
```

Both are loops that can go up to infinity. But I do not know which one of them is clearer.

Comment (dhruvmanila) [reply searches/fibonacci_search.py]: As far as I understand, you are finding the index of a fibonacci number which is greater than the length of the input array, right? If that's the case then why are you using a generator just to find one value?

It seems like I probably found a bug:
```python
print(fibonacci_search(list(range(100)), 63))

...
Traceback (most recent call last):
  File "fibonacci_search.py", line 117, in <module>
    print(fibonacci_search(list(range(100)), 63))
  File "fibonacci_search.py", line 106, in fibonacci_search
    offset += fibonacci(fibb_k - 2)
  File "fibonacci_search.py", line 49, in fibonacci
    raise ValueError("k integer must be greater or equal to zero.")
ValueError: k integer must be greater or equal to zero.
```

Comment (NumberPiOso) [reply searches/fibonacci_search.py]: > As far as I understand, you are finding the index of a fibonacci number which is greater than the length of the input array, right? If that's the case then why are you using a generator just to find one value?

Because when I use, next of a generator it will only find the first value of the iterator. And the first value of that generator is the value which I am looking for.

I think is OK but the code in this repo should be clean and understandable. Do you think that it is better for me to change it to something easier? like the snippet above.

Comment (dhruvmanila) [reply searches/fibonacci_search.py]: I think the above mentioned snippet is more readable (with descriptive variable names). Also, take a look at the error produced in the above comment.

Comment (NumberPiOso) [reply searches/fibonacci_search.py]: > I think the above mentioned snippet is more readable (with descriptive variable names).
Agreed, I change it for the sake of clarity.

>  Also, take a look at the error produced in the above comment.
Thank you for pointing that out, it is a special case of the fibonacci search. I fixed it and added your example and many more with big list as doctests.

**Comment by anilbaday** [comment]
What is this ?

iOS için Outlook<https://aka.ms/o0ukef> uygulamasını edinin
________________________________
Gönderen: Dhruv <notifications@github.com>
Gönderildi: Wednesday, October 7, 2020 5:28:49 PM
Kime: TheAlgorithms/Python <Python@noreply.github.com>
Bilgi: Subscribed <subscribed@noreply.github.com>
Konu: Re: [TheAlgorithms/Python] Fix: Multiple errors in fibonacci search. (#2659)


@dhruvmanila commented on this pull request.

________________________________

In searches/fibonacci_search.py<https://github.com/TheAlgorithms/Python/pull/2659#discussion_r501056383>:

> -            fibNext = fib_N_1
-            fib_N_1 = fib_N_2
-            fib_N_2 = fibNext - fib_N_1
-            index = i
-        elif arr[i] > val:
-            fibNext = fib_N_2
-            fib_N_1 = fib_N_1 - fib_N_2
-            fib_N_2 = fibNext - fib_N_1
-        else:
-            return i
-    if (fib_N_1 and index < length - 1) and (arr[index + 1] == val):
-        return index + 1
-    return -1
+    n = len(arr)
+    # Find m such that F_m >= n where F_i is the i_th fibonacci number.
+    m = next(x for x in range(sys.maxsize ** 10) if fibonacci(x) >= n)


As far as I understand, you are finding the index of a fibonacci number which is greater than the length of the input array, right? If that's the case then why are you using a generator just to find one value?

It seems like I probably found a bug:

print(fibonacci_search(list(range(100)), 63))

...
Traceback (most recent call last):
  File "fibonacci_search.py", line 117, in <module>
    print(fibonacci_search(list(range(100)), 63))
  File "fibonacci_search.py", line 106, in fibonacci_search
    offset += fibonacci(fibb_k - 2)
  File "fibonacci_search.py", line 49, in fibonacci
    raise ValueError("k integer must be greater or equal to zero.")
ValueError: k integer must be greater or equal to zero.

File: searches/fibonacci_search.py [hunk 10]
```diff
@@ -1,51 +1,130 @@
...
-    return -1
+    len_list = len(arr)
+    # Find m such that F_m >= n where F_i is the i_th fibonacci number.
+    i = 0
+    while True:
+        if fibonacci(i) >= len_list:
+            greater_fibb_index = i
+            break
+        i += 1
+    fibb_k = greater_fibb_index
```
Comment (dhruvmanila) [inline searches/fibonacci_search.py:112]:
```suggestion
            fibb_k = i
            break
        i += 1
```
You can directly assign it in the loop.

**Review by dhruvmanila — APPROVED** [review]
Looks good now! Thank you for your contribution!
